File: src/balances.ts
import type { Exchange } from "ccxt";

export async function gateioSpotBalance(spot: Exchange, coin: string): Promise<number> {
  try {
    const bal: any = await spot.fetchBalance();
    const entry: any = bal?.[coin.toUpperCase()] ?? {};
    const free = Number(entry?.free ?? 0);
    const total = Number(entry?.total ?? 0);
    return total > 0 ? total : free;
  } catch {
    return 0.0;
  }
}

export async function bithumbSpotBalance(bithumb: Exchange, coin: string): Promise<number> {
  try {
    const bal: any = await bithumb.fetchBalance();
    const entry: any = bal?.[coin.toUpperCase()] ?? {};
    const free = Number(entry?.free ?? 0);
    const total = Number(entry?.total ?? 0);
    return total > 0 ? total : free;
  } catch {
    return 0.0;
  }
}

File: legacy_py/arbitrage/balances.py
"""Balance helpers."""

from __future__ import annotations

import ccxt


def gateio_spot_balance(spot: ccxt.Exchange, coin: str) -> float:
    try:
        bal = spot.fetch_balance()
        entry = bal.get(coin.upper(), {}) if isinstance(bal, dict) else {}
        free = float(entry.get("free") or 0.0)
        total = float(entry.get("total") or 0.0)
        return total if total > 0 else free
    except Exception:
        return 0.0


def bithumb_spot_balance(bithumb: ccxt.Exchange, coin: str) -> float:
    try:
        bal = bithumb.fetch_balance()
        entry = bal.get(coin.upper(), {}) if isinstance(bal, dict) else {}
        free = float(entry.get("free") or 0.0)
        total = float(entry.get("total") or 0.0)
        return total if total > 0 else free
    except Exception:
        return 0.0
